smart-contracts/assembly/contracts/NFT/NFTEnumerable-example.ts

```diff
@@ -0,0 +1,253 @@
+ *
+ * We store important information in the following formats:
+ *
+ * - **Total Supply:**
+ *
+ *   [TOTAL_SUPPLY_KEY] = totalSupply
+ *   - `TOTAL_SUPPLY_KEY`: A constant key for the total supply of tokens.
+ *   - `totalSupply`: A `u256` value representing the total number of tokens in existence.
+ *
+ * - **Owned Tokens:**
+ *
+ *   [OWNED_TOKENS_KEY][owner][tokenId] = tokenId
+ *   - `OWNED_TOKENS_KEY`: A constant prefix for all owned tokens.
+ *   - `owner`: The owner's address.
+ *   - `tokenId`: The token ID.
+ *   - The value `tokenId` is stored to facilitate easy retrieval.
+ *
+ * **Retrieving Data Using Key Prefixes:**
+ *
+ * We utilize the `getKeys` function from the `massa-as-sdk`, which allows us to retrieve all keys that start with a
+ * specific prefix. This enables us to:
+ * - Retrieve all tokens owned by a specific address by querying keys with the prefix `[OWNED_TOKENS_KEY][owner]`.
+ * - Retrieve all existing token IDs by querying keys with the appropriate prefix if needed.
+ *
+ * **Key Points:**
+ * - The `getKeys` function from the `massa-as-sdk` allows us to filter storage keys by a given prefix,
+ *  enabling efficient data retrieval.
+ *
+ * **This file does two things:**
+ * 1. It wraps the `NFTEnumerable-internals` functions, manages the deserialization/serialization of the arguments
+ *    and return values, and exposes them to the outside world.
+ * 2. It implements some custom features that are not part of the ERC-721 standard,
+ *    such as `mint`, `burn`, or ownership management.
+ *
+ * **Important:** The `NFTEnumerable-internals` functions are not supposed to be re-exported by this file.
+ */
+
+import {
+  Args,
+  boolToByte,
+  stringToBytes,
+  u256ToBytes,
+} from '@massalabs/as-types';
+import {
+  _approve,
+  _balanceOf,
+  _constructor,
+  _getApproved,
+  _isApprovedForAll,
+  _name,
+  _ownerOf,
+  _setApprovalForAll,
+  _symbol,
+  _update,
+  _transferFrom,
+  _totalSupply,
+} from './NFTEnumerable-internals';
+import { setOwner, onlyOwner } from '../utils/ownership';
+import { Context, isDeployingContract } from '@massalabs/massa-as-sdk';
+
+/**
+ * @param binaryArgs - serialized strings representing the name and the symbol of the NFT
+ *
+ * @remarks This is the constructor of the contract. It can only be called once, when the contract is being deployed.
+ * It expects two serialized arguments: the name and the symbol of the NFT.
+ * Once the constructor has handled the deserialization of the arguments,
+ * it calls the _constructor function from the NFT-enumerable-internals.
+ *
+ * Finally, it sets the owner of the contract to the caller of the constructor.
+ */
+export function constructor(binaryArgs: StaticArray<u8>): void {
+  assert(isDeployingContract());
+  const args = new Args(binaryArgs);
+  const name = args.nextString().expect('name argument is missing or invalid');
+  const symbol = args
+    .nextString()
+    .expect('symbol argument is missing or invalid');
```

Comment:
I think "name" and "symbol" should be hardcoded here. there is no real need of passing it with args.. and it would simplify the example

Reply:
Sure thanks